packages/apollo-federation/src/composition/validate/postComposition/keySelectionSetsDeclared.ts

```diff
@@ -0,0 +1,91 @@
+import { GraphQLSchema, isObjectType, FieldNode, GraphQLError } from 'graphql';
+
+import { logServiceAndType, errorWithCode } from '../../utils';
+
+/**
+ * - Each key's selection set must be present on the base service
+ */
+export const keySelectionSetsDeclared = (schema: GraphQLSchema) => {
+  const errors: GraphQLError[] = [];
+
+  const types = schema.getTypeMap();
+  for (const [typeName, namedType] of Object.entries(types)) {
+    if (!isObjectType(namedType)) continue;
+
+    if (namedType.federation && namedType.federation.keys) {
+      const baseService = namedType.federation.serviceName!;
+      if (!baseService) {
+        // Federated type does not have a base service. This shouldn't happen, but does in some tests.
```

Comment:
packages/apollo-federation/src/composition/__tests__/composeAndValidate.test.ts
  ● composition of value types › errors › when used as an entity

Reply:
```
      const serviceA = {
        typeDefs: gql`
          type Query {
            product: Product
          }

          type Product {
            sku: ID!
            color: String!
          }
        `,
        name: 'serviceA',
      };

      const serviceB = {
        typeDefs: gql`
          type Query {
            topProducts: [Product]
          }

          type Product @key(fields: "sku") {
            sku: ID!
            color: String!
          }
        `,
        name: 'serviceB',
      };
```

Reply:
Ah, that's helpful and matches my expectation. This is a case you don't have to account for (as it seems you're already aware), since it's handled in another validation 👍